### PR TITLE
Support for special variables

### LINKED ISSUE #16: Option for note title in template
Add an option to specify the title of the note in the template, including the possibility to use variables.

### PR DESCRIPTION
**Pull request by nishantwrp**
Fixes #16

### CODE REVIEW

**Comment by CalebJohn** [comment]
@nishantwrp The code looks good, but I found a bug related to parsing.

```
---
var1: enum(a, b, c)
template_tags: {{var1}}
---

Some body
```
image
Same error for 
```
---
template_title: {{date}}
---

Some body
```

I also have a suggestion, if the user inserts a template in to a note, the tags should still be applied. This will make the plugin useful as both a note body template plugin, but also a note tagging template plugin.

**Comment by nishantwrp** [comment]
Actually in that case the the variable needs to be wrapped in double quotes. 

template_tags: "{{ var1 }}"

I'll maybe add a line in documentation to always wrap special variables in double quotes. Thoughts?

**Comment by nishantwrp** [comment]
> I also have a suggestion, if the user inserts a template in to a note, the tags should still be applied. This will make the plugin useful as both a note body template plugin, but also a note tagging template plugin.

This is a nice idea indeed, but maybe there are some users who don't want the tags when the insert template is used. Personally, I may not want it to.

**Comment by CalebJohn** [comment]
> I'll maybe add a line in documentation to always wrap special variables in double quotes. Thoughts?

I think this is a situation where you should look for a technical solution. Keep in mind that the majority of your users won't be programmers, they won't know/care what YAML is. I'll be very surprised if this is an issue that you can't overcome with a couple lines of code.

> This is a nice idea indeed, but maybe there are some users who don't want the tags when the insert template is used. Personally, I may not want it to.

It's a bit of an awkward issue isn't it? Some users will likely expect the tags (like me), and others will be shocked to see them. It makes sense to save yourself the effort until it becomes clear that users want it.

**Comment by nishantwrp** [comment]
>  I'll be very surprised if this is an issue that you can't overcome with a couple lines of code.

A solution that I can think of is auto wrap special variable values with double quotes if they are not already.

**Comment by nishantwrp** [comment]
@CalebJohn done. Please re-review.

File: src/parser.ts [hunk 1]
```diff
@@ -151,11 +151,79 @@ export class Parser {
...
+                return null;
+            }
+
+            const allMatchesAfterFirstMatch = allLines.map((val, index) => {
+                if (index && val === "---") {
+                    return index;
+                } else {
+                    return null;
+                }
+            }).filter(val => !!val);
```
Comment (CalebJohn) [inline src/parser.ts:175]: Clever solution! I think a for loop would be faster, but it doesn't matter here.

File: src/parser.ts [hunk 2]
```diff
@@ -151,11 +151,79 @@ export class Parser {
...
+                return null;
+            }
+
+            const blockEndIndex = allMatchesAfterFirstMatch[0];
+            return allLines.slice(0, blockEndIndex + 1).join("\n");
+        }
+
+        const wrapInQuotes = (definitionsBlock: string, properties: string[]) => {
+            for (const prop of properties) {
+                const pattern = new RegExp(`^ *${prop}:.*`, "gm");
```
Comment (CalebJohn) [inline src/parser.ts:187]: You should use `[^\S\n]` rather than a literal space here. It will allow you to catch all (non-newline) whitespace.

Comment (nishantwrp) [reply src/parser.ts]: done

**Review by CalebJohn — APPROVED** [review]

**Comment by CalebJohn** [comment]
LGTM!

**Pull request merged**